Is obstructed method

Edwin-F opened the pull request:
This method is nested under the piece model. I also wrote rpsec tests for the four conditions. You can run the 4 tests using:
$ bundle exec rspec spec/models/game_spec.rb:24

this runs the rspec file in spec/models/game_spec.rb  for block of code starting at line 24

or run all the tests
$ bundle exec rspec spec

or try the command in the console
Game.create(id: 1, white_player_id: 1, black_player_id: 2)
Piece.last.is_obstructed?(7, 0)

FYI, I started writing rspec tests on the model and controller to test functionally since I was kinda lost on what was approved in the past merges and if my new method broke a previous function. For future methods and functionally, I suggest writing rspec tests to make it easier to keep track of functions. Thanks!

Edwin-F committed:
start rspect tests for is_obstructed method

Edwin-F committed:
expanded obstructed method and rspec tests

Edwin-F committed:
fixed bug for last test

Edwin-F committed:
refactored method

xelzi commented:
Nice job @Edwin-F .... two things

1) I tried to do a little refactoring. I don't think you need both an array_down and array_up function....I combined these functions into one function called locations array that uses an if statement to construct the array...that simplifies a lot of the stuff in the individual cases. I changed this code on my local branch and all the tests work except one part of the horizontal obstruction...will try to figure that out



`def is_obstructed?(row_final, col_final)
    row_init = self.position_row
    col_init = self.position_column

    def locations_array(init, final)
      if init < final
        return (init + 1...final).to_a
      else
        return (init-1).downto(final + 1).to_a
      end
    end

    #horizonal case
    if row_init == row_final
      return !Piece.exists?(position_row: row_init, position_column: locations_array(row_init, row_final))

      #vertical case
    elsif col_init == col_final 
      return !Piece.exists?(position_row: locations_array(row_init, row_final), position_column: col_init)

      #diagonal case
    elsif ((row_final - row_init).to_f/(col_final - col_init).to_f ).abs == 1
      return !Piece.exists?(position_row: locations_array(row_init, row_final), position_column: locations_array(col_init, col_final))
    
      #invalid input case
    else
      raise RuntimeError, "invalid input. Not diagonal, horizontal, or vertical."
    end
  end`

2) If I am understanding correctly, is_obstructed? is returning false if obstruction exists. Shouldn't we have it the other around? If we call is_obstructed? we expect true if there is an obstruction and false if there is not an obstruction.

The tests look really good, I am going to use them as a guide going forward.

smiths15 commented:
@Edwin-F  I will look at the code and comments mentioned above. I also noticed there is a merge conflict, it seems like the capture logic wasn't merged into this branch.

smiths15 commented on app/models/piece.rb:
```diff
@@ -32,4 +32,57 @@ def valid_move?
+      else #going left
+        col_rng = array_down(col_init, col_final)
+      end
+      return !Piece.exists?(position_row: row_init, position_column: col_rng)
```
@Edwin-F  I think @xelzi is right, when we call is_obstructed, the method should return true if there is a piece in the way, should return false if a piece is not in the way, and should raise an error if the piece doesn't move horizontal, vertical, or diagonal (i.e., a knight)

smiths15 commented on spec/models/game_spec.rb:
```diff
@@ -0,0 +1,65 @@
+  end
+
+
+  describe "game pieces method is_Obstructed?" do
```
The test look good and should be updated to match the logic mentioned above for the is_obstructed method.
The test related to the piece model should be created in a separate spec file --> "piece_spec.rb"

smiths15 reviewed (changes requested):
Great work! See comments for suggested changes.

Edwin-F committed:
refactor method after review and added piece_spec.rb

Edwin-F commented:
Thanks for the suggestions all. I refactored the code, updated the rspec tests for true, and added the piece_spec.rb. Let me know if there are other suggestions.
you can test the rspec tests for the piece model here with this command
bundle exec rspec spec/models/piece_spec.rb

and game model with this
bundle exec rspec spec/models/game_spec.rb

smiths15 commented on app/models/piece.rb:
```diff
@@ -32,4 +32,36 @@ def valid_move?
+    row_range = array_position(row_init, row_final)
+    col_range = array_position(col_init, col_final)
+
+    #horizonal case
```
Works great 👍

smiths15 commented on app/models/piece.rb:
```diff
@@ -32,4 +32,36 @@ def valid_move?
+    if row_init == row_final
+      return Piece.exists?(position_row: row_init, position_column: col_range)
+
+      #vertical case
```
works great 👍

smiths15 commented on app/models/piece.rb:
```diff
@@ -32,4 +32,36 @@ def valid_move?
+    end
+
+    #invalid input case
+    raise RuntimeError, "invalid input. Not diagnal, horizontal, or vertical."
```
works great 👍

smiths15 commented on app/models/piece.rb:
```diff
@@ -32,4 +32,36 @@ def valid_move?
+    elsif col_init == col_final 
+      return Piece.exists?(position_row: row_range, position_column: col_init)
+
+      #diagonal case
```
works great 👍

smiths15 commented on spec/models/piece_spec.rb:
```diff
@@ -0,0 +1,44 @@
+require 'rails_helper'
```
The tests all pass 👍 

I would also suggest testing the cases where false is also true, which should be a simple add

smiths15 committed:
Merge branch 'master' into Is_Obstructed_Method

smiths15 commented:
@Edwin-F I fixed the merge conflict to include the capture_logic branch

xelzi commented:
@smiths15 I don't know what happened, but when I pull down the code now, I get 7 failures...I deleted the branch and pulled it down again and have the same problem. Would you confirm that all the tests are working for you with the current code after resolving the merge conflict (attached is a picture of my errors)

<img width="866" alt="screenshot 2017-06-20 08 45 41" src="https://user-images.githubusercontent.com/14365312/27333751-f810b6ae-5594-11e7-8735-43475da1f3c8.png">

xelzi commented:
@Edwin-F @smiths15  
So the problem isn't with Edwin's code...the problem is with the main code (I think).
Edwin's tests just catch some of those problems (nice job Edwin haha). @smiths15 did you merge the master into Edwin's code for testing? If not, it makes sense why the tests worked for you and not me..thats all I can think of at least....

To see what I mean, you can follow the command line instructions
git fetch origin
git checkout -b Is_Obstructed_Method origin/Is_Obstructed_Method
git merge master


For example, one of the errors I am getting arrises when testing if we have 32 game pieces. I went to the rails console in the main code and saw that if I created a game, no pieces were automatically created there either. That is why I think that these errors stems from the master branch and not this local one...

smiths15 commented:
@xelzi I'm looking into the errors now

smiths15 commented:
@xelzi and @Edwin-F, if you call @game.set_pieces_on_board in the spec files it should work

```
before do
    @game = Game.create(id: 1, white_player_id: 1, black_player_id: 2)
    @game.set_pieces_on_board
  end
```

smiths15 commented on spec/models/game_spec.rb:
```diff
@@ -0,0 +1,30 @@
+RSpec.describe Game, type: :model do
+  before do
+    @game = Game.create(id: 1, white_player_id: 1, black_player_id: 2)
+  end
```
add the following to the spec file so the rspec tests will pass
```
@game.set_pieces_on_board
```

smiths15 commented on spec/models/piece_spec.rb:
```diff
@@ -0,0 +1,44 @@
+RSpec.describe Piece, type: :model do
+  before do
+    @game = Game.create(id: 1, white_player_id: 1, black_player_id: 2)
+  end
```
add the following so the rspec tests will pass
```
@game.set_pieces_on_board
```

smiths15 commented on spec/models/game_spec.rb:
```diff
@@ -0,0 +1,30 @@
+    it 'has 32 pieces' do
+      piece_count = Piece.count
+      game_count = Game.count
+      #Game.last.pieces.each {|piece| p piece}
```
not necessary

smiths15 commented on spec/models/game_spec.rb:
```diff
@@ -0,0 +1,30 @@
+    end
+
+    it 'has 16 white pieces & 16 black pieces' do
+      game_count = Game.count
```
not necessary since it was called earlier

smiths15 commented on spec/models/game_spec.rb:
```diff
@@ -0,0 +1,30 @@
+      game_count = Game.count
+      white_piece_count = Piece.where(:color => 'white').count
+      black_piece_count = Piece.where(:color => 'black').count
+      expect(game_count).to eq 1
```
can also remove this

smiths15 commented on spec/models/game_spec.rb:
```diff
@@ -0,0 +1,30 @@
+    end
+
+    it 'begins with white player\'s turn' do
+      expect(@game.turn).to eq 1
```
add @game.set_default_turn so this test passes

xelzi commented:
@smiths15 nice detective work. Glad you got it to work. My only concern is that Game.create should generate pieces on the board already since the create action in the games controller calls set pieces on the board. I would say this is only an rspec issue, except for the fact that when I create a game in the console, the pieces are also not set on the board...and that was not the case previously.....I am still thinking about what the issue might be.

xelzi commented:
EDIT: Sorry for posting so much....I actually understand the problem now and think we should just go ahead with Smith's suggestion of adding the game.set_piece_on_board to the tests to solve the problem

Edwin-F committed:
after merge with master, fixed rspec tests from review

Edwin-F commented:
I pull the most recent commit that was merged with the master branch and updated the rspec tests to make them a success again. Let me know if further changes are needed. Thanks!

FYI, if you ever need to debug something related to a method or variable, adding a line of code in the rspec file or within the method itself
puts variable_name 
or 
p variable
will print it on the command line when you run the command bundle exec rspec.

xelzi commented:
Hi! Most of the tests are fixed...there is still one problem for me...probably an easy fix! I don't have too much time right now but will have more time this weekend
<img width="556" alt="screenshot 2017-06-22 18 22 23" src="https://user-images.githubusercontent.com/14365312/27458599-d1d56270-5777-11e7-91af-3a047c01a3b2.png">

Edwin-F commented:
Can you check if more than one game was created on your system (16x2=32)? 
The tests come up successful on my computer.

![image](https://user-images.githubusercontent.com/25597947/27504614-08de85c2-5853-11e7-93f7-e960630e94e4.png)

xelzi commented:
@Edwin-F I just tried again and still got the same error. No idea why, when I create a game in the rails console I definitely get 16 white and 16 black pieces. I see no errors with your test. @smiths15 @JefeNelson @SomeRandom42  if any of you can test Edwin's code without error, I think it should be fine. I will continue to troubleshoot my own problem.

xelzi commented:
@Edwin-F we are all good now! Error on my end, thanks!
